### PR TITLE
Make transaction message compilation consistent with @solana/web3.js

### PR DESCRIPTION
**Pull request by abacus-x**
### Note
Hi, I noticed an issue when serializing transactions - identical transactions had different compiled message bytes using JavaScript and Python. The primary issue stemmed from differences in sorting AccountMetas, which can result in different account index references and serialized bytes. The JavaScript package includes a third sort key that references the AccountMeta PublicKey (https://github.com/solana-labs/solana-web3.js/blob/master/src/transaction.ts, also https://docs.rs/solana-program/1.6.8/src/solana_program/message.rs.html). 

### Changes

Three primary changes:
1. Updated AccountMeta sorting to match solana/web3.js package. The sorting was a bit different, which resulted in different account key indexes and serialized messages (identical transaction had different serialized output vs js package). The js package has a third sort key that references the AccountMeta PublicKey.
2. Updated the way signatures are added. Previously it added signer signatures if none were present - now it iterates through signers to see if any are missing.
3. Simplified the way AccountMetas are culled

### CODE REVIEW

**Comment by michaelhly** [comment]
@abacus-x this is great. Would you mind making sure all the CI check passes, and write a test case on presenting that re-produces the issue in the current version, but passes after your changes?

**Comment by abacus-x** [comment]
@michaelhly  sure, I'll submit both items this upcoming week

**Comment by abacus-x** [comment]
@michaelhly, as promised, the PR now includes the test you requested.  

I want to highlight a few points for consideration before we consider merging:
1. Solana CLI serialization does not include the third sort key (PublicKey). As such, it serializes the same way as solana-py did prior to this PR. JavaScript does implement the sort key (https://github.com/solana-labs/solana-web3.js/blob/master/src/transaction.ts). Transactions currently do not serialize/deserialize properly between Python and JavaScript.
2. We should consider notifying the other package maintainers of the issue so a common solution can be reached – this would address cross-package serialization/deserialization errors. Please let me know your thoughts.
3. I created a simple JavaScript test to help walk you through the derivation of the base64 strings used in Python tests (https://github.com/abacus-x/solana-tx-serialization).
4. Given the CLI/JS discrepancies, I modified the Python tests that reference base64 strings (test_transaction.py, test_vote_program.py, test_system_program.py) – they now include both CLI and JS base64 strings. We could tweak this method based on your preference.  

I hope this is helpful, and let me know if you require additional clarification.

File: tests/unit/test_system_program.py
Comment (michaelhly) [inline tests/unit/test_system_program.py:243]:
```suggestion
    assert create_account_txn == js_expected_txn
    # XXX:  Cli message serialization do not sort on account metas producing discrepency 
    # assert create_account_txn == cli_expected_txn
```

File: tests/unit/test_vote_program.py
Comment (michaelhly) [inline tests/unit/test_vote_program.py:113]:
```suggestion
     assert serialized_message == js_wire_msg
    # XXX:  Cli message serialization do not sort on account metas producing discrepency 
    # serialized_message txn == cli_wire_msg
```

**Review by michaelhly — COMMENTED** [review]
A few comments:

- Let's try to avoid `str(pubkey)` as much as possible b/c it does an underlying base58 decode and slows down the serialization process:
https://github.com/michaelhly/solana-py/blob/f527e129c5d4f5d7218e452e1b27a15b96c0f55e/src/solana/publickey.py#L68-L70
- Instead of doing `assert create_account_txn in [js_expected_txn, cli_expected_txn]` let's  be clear here and comment on why the other one does not match.

File: src/solana/transaction.py
Comment (abacus-x) [reply src/solana/transaction.py:203]: Yes, makes sense, that is a better way to sort it and reduce the use of str(PublicKey)

File: src/solana/transaction.py
Comment (abacus-x) [reply src/solana/transaction.py:231]: LGTM

File: src/solana/transaction.py
Comment (abacus-x) [reply src/solana/transaction.py]: Yes, just added this request

**Comment by abacus-x** [comment]
@michaelhly an issue occurred with my local dev environment (in event you are looking at it now).  One moment please.

File: tests/unit/test_transaction.py
Comment (michaelhly) [inline tests/unit/test_transaction.py:17]:
```suggestion
```

File: tests/unit/test_vote_program.py
Comment (michaelhly) [inline tests/unit/test_vote_program.py:115]:
```suggestion
```

File: src/solana/transaction.py
Comment (michaelhly) [inline src/solana/transaction.py:251]:
```suggestion
                account_keys=[k for (k, _) in account_keys],
```

**Review by michaelhly — APPROVED** [review]
@abacus-x thank you! Will merge after all the CI checks pass!

**Comment by abacus-x** [comment]
> @abacus-x thank you! Will merge after all the CI checks pass!

@michaelhly  awesome, happy to help - great and useful library

**Pull request merged**
